src/library_debugger_toolkit.js

```diff
@@ -0,0 +1,477 @@
+    "u8" : 0,
+    "u16": 1,
+    "u32": 2,
+    "i8" : 0,
+    "i16": 1,
+    "i32": 2,
+    "f32": 2,
+    "f64": 3
+  }
+  function install_heap($heap) {
+    heap = {
+      "u8" : new Uint8Array($heap),
+      "u16": new Uint16Array($heap),
+      "u32": new Uint32Array($heap),
+      "i8" : new Int8Array($heap),
+      "i16": new Int16Array($heap),
+      "i32": new Int32Array($heap),
+      "f32": new Float32Array($heap),
+      "f64": new Float64Array($heap)
+    };
+  }
+
+  var symbols = {};
+  function stackTrace() {
+      var err = new Error();
+      return err.stack;
+  }
+
+  function install_cyberdwarf(data_cd) {
+    cyberdwarf = JSON.parse(data_cd)["cyberdwarf"];
+    invert_vtables();
+  }
+
+  function type_descriptor_to_heap_id(type_descriptor) {
+    var id = "";
+    switch (type_descriptor[TAG_IDX]) {
+      case DW_ATE_unsigned: case DW_ATE_unsigned_char: id = "u"; break;
+      case DW_ATE_signed: case DW_ATE_signed_char: id = "i"; break;
+      case DW_ATE_float: id = "f"; break;
+    }
+    id += type_descriptor[4];
+    return id;
+  }
+
+  function type_id_to_type_descriptor(type_id, ptr, dit) {
+    if (!isNaN(+type_id)) {
+      return cyberdwarf["types"][type_id];
+    }
+    if (typeof(type_id) === "string") {
+      if (dit) {
+        type_id = resolve_from_vtable(ptr, type_id);
+      }
+      type_id = cyberdwarf["type_name_map"][type_id];
+    }
+    return cyberdwarf["types"][type_id];
+  }
+
+  function invert_vtables() {
+    cyberdwarf["has_vtable"] = {};
+    for (var i in cyberdwarf["vtable_offsets"]) {
+      cyberdwarf["has_vtable"][cyberdwarf["vtable_offsets"][i]] = true;
+    }
+  }
+
+  function resolve_from_vtable(val, type_name) {
+    var lookup_name = "_ZTV" + type_name.substr(4);
+    if (cyberdwarf["has_vtable"][lookup_name]) {
+      var potential_vtable = "" + (heap["u32"][val >> 2] - 8);
+      if (cyberdwarf["vtable_offsets"][potential_vtable]) {
+        var ans = type_name.substr(0,4) + cyberdwarf["vtable_offsets"][potential_vtable].substr(4);
+        return ans;
+      }
+    }
+    return type_name;
+  }
+
+  var current_function = "";
+  function set_current_function(func_name) {
+    if (typeof(cyberdwarf["function_name_map"][func_name]) !== "undefined") {
+      func_name = cyberdwarf["function_name_map"][func_name];
+    }
+    if (func_name.substring(0,1) == "_") {
+      func_name = func_name.substring(1);
+    }
+    current_function = cyberdwarf["functions"][func_name];
+  }
+
+  function pretty_print_to_object(val, type_id, depth) {
+    install_heap(Module.HEAPU8.buffer);
+    var base_type = new TypedVariable(val, current_function[type_id]);
+    return base_type.prettyPrintToObject(depth);
+  }
+
+  function pretty_print_from_typename(val, type_name, depth) {
+    install_heap(Module.HEAPU8.buffer);
+    var base_type = new TypedVariable(val, type_name);
+    return base_type.prettyPrintToObject(depth);
+  }
+
+  function stack_decoder(val, name, depth, on_heap) {
```

Comment:
Used jsStackTrace. In this case we want mangled names to look up the right type. Instead of having the user know the mangled type name, this lets them just know the Javascript function name and the variable name.